# Market Metrics Calculation

In [1]:
# Dependencies
import pandas as pd
import numpy as py
import matplotlib.pyplot as plt
import seaborn as sns

# ignore the warning
from warnings import filterwarnings
filterwarnings("ignore")

In [2]:
#File path
footfall_data_path = "footfall_uk.csv"
retail_data_path = "retail_clean.csv"

# Load the data
footfall_data = pd.read_csv(footfall_data_path, encoding="utf-8")
retail_data = pd.read_csv(retail_data_path, encoding="utf-8")

# Review
# footfall_data
retail_data

Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041          RECORD FRAME 7" SINGLE SIZE        48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1042716  581587     22899          CHILDREN'S APRON DOLLY GIRL         6   
1042717  581587     23254         CHILDRENS CUTLERY DOLLY GIRL         4   
1042718  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1042719  581587     22138         BAKING SET 9 PIECE RETROSPOT         3   
1042720  581587      POST                              POSTAGE         1   

                  InvoiceDate  Price  Customer ID         Country  
0        2009-12-01T07:45:00Z   6.95      13085.0  United Kingdom  
1        2009-12-01T07:45:00Z   6.75      13085.0  United Kingdom  
2        2009-12-01T07:45:00Z   6.75      13085.0  United Kingdom  
3        2009-12-01T07:45:00Z   2.10      13085.0  United Kingdom  
4        2009-12-01T07:45:00Z   1.25      13085.0  United Kingdom  
...                       ...    ...          ...             ...  
1042716  2011-12-09T12:50:00Z   2.10      12680.0          France  
1042717  2011-12-09T12:50:00Z   4.15      12680.0          France  
1042718  2011-12-09T12:50:00Z   4.15      12680.0          France  
1042719  2011-12-09T12:50:00Z   4.95      12680.0          France  
1042720  2011-12-09T12:50:00Z  18.00      12680.0          France  

[1042721 rows x 8 columns]

In [3]:
footfall_data

no                 date  website_visitors
0      0  2009-12-01 00:00:00               471
1      1  2009-12-02 00:00:00               456
2      2  2009-12-03 00:00:00               491
3      3  2009-12-04 00:00:00               353
4      4  2009-12-05 00:00:00               119
..   ...                  ...               ...
599  599  2011-12-05 00:00:00               515
600  600  2011-12-06 00:00:00               456
601  601  2011-12-07 00:00:00               428
602  602  2011-12-08 00:00:00               479
603  603  2011-12-09 00:00:00               170

[604 rows x 3 columns]

In [4]:
# check info 
footfall_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 604 entries, 0 to 603
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   no                604 non-null    int64 
 1   date              604 non-null    object
 2   website_visitors  604 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 14.3+ KB


In [5]:
footfall_data["date"] = pd.to_datetime(footfall_data["date"])

footfall_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 604 entries, 0 to 603
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   no                604 non-null    int64         
 1   date              604 non-null    datetime64[ns]
 2   website_visitors  604 non-null    int64         
dtypes: datetime64[ns](1), int64(2)
memory usage: 14.3 KB


In [6]:
retail_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1042721 entries, 0 to 1042720
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1042721 non-null  object 
 1   StockCode    1042721 non-null  object 
 2   Description  1042721 non-null  object 
 3   Quantity     1042721 non-null  int64  
 4   InvoiceDate  1042721 non-null  object 
 5   Price        1042721 non-null  float64
 6   Customer ID  805618 non-null   float64
 7   Country      1042721 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 63.6+ MB


In [7]:
retail_data["InvoiceDate"] = pd.to_datetime(retail_data["InvoiceDate"])

retail_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1042721 entries, 0 to 1042720
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype              
---  ------       --------------    -----              
 0   Invoice      1042721 non-null  object             
 1   StockCode    1042721 non-null  object             
 2   Description  1042721 non-null  object             
 3   Quantity     1042721 non-null  int64              
 4   InvoiceDate  1042721 non-null  datetime64[ns, UTC]
 5   Price        1042721 non-null  float64            
 6   Customer ID  805618 non-null   float64            
 7   Country      1042721 non-null  object             
dtypes: datetime64[ns, UTC](1), float64(2), int64(1), object(4)
memory usage: 63.6+ MB


## 1.which week year has the highest conversion rate?

In [18]:
# Generate new column: week
retail_data["week"] = retail_data["InvoiceDate"].dt.to_period("W").apply(lambda r: r.start_time)

# Calculate the cnt_cutomer in retail
sum_customer = (
    retail_data.groupby("week")["Customer ID"]
    .nunique()   # count distinct
    .reset_index(name="cnt_customer")
    .sort_index()
)

sum_customer

week  cnt_customer
0   2009-11-30           427
1   2009-12-07           411
2   2009-12-14           359
3   2009-12-21            54
4   2010-01-04           167
..         ...           ...
99  2011-11-07           517
100 2011-11-14           569
101 2011-11-21           493
102 2011-11-28           516
103 2011-12-05           415

[104 rows x 2 columns]

In [19]:
footfall_data["week"] = footfall_data["date"].dt.to_period("W").apply(lambda r: r.start_time)

visitor_per_week = (
    footfall_data.groupby("week")["website_visitors"]
    .sum()
    .reset_index(name="sum_visitor")
    .sort_index()
)

visitor_per_week

week  sum_visitor
0   2009-11-30         2183
1   2009-12-07         2103
2   2009-12-14         1949
3   2009-12-21          427
4   2010-01-04          812
..         ...          ...
99  2011-11-07         2551
100 2011-11-14         2790
101 2011-11-21         2517
102 2011-11-28         2670
103 2011-12-05         2048

[104 rows x 2 columns]

In [25]:
# left join dateset
joined_data = pd.merge(sum_customer, visitor_per_week, how="left", on="week")

#Calculate conversation rate
joined_data["conversion_rate"] = round(joined_data["cnt_customer"] / joined_data["sum_visitor"] * 100, 2)

# sort by DESC
joined_data = joined_data.sort_values(by="conversion_rate", ascending=False)

# dispaly
joined_data

week  cnt_customer  sum_visitor  conversion_rate
85 2011-08-01           275         1248            22.04
87 2011-08-15           277         1264            21.91
55 2011-01-03           207          959            21.58
62 2011-02-21           245         1150            21.30
63 2011-02-28           242         1137            21.28
..        ...           ...          ...              ...
80 2011-06-27           205         1144            17.92
16 2010-03-29           214         1224            17.48
41 2010-09-20           323         1930            16.74
54 2010-12-20            86          642            13.40
3  2009-12-21            54          427            12.65

[104 rows x 4 columns]

### Answer
The week starts at Aug 1, 2011 has highest conversion rate.

## 2. Which week year has  highest average transation value?

In [ ]:
# Genarate value per transation
retail_data["value"] = retail_data["Quantity"] * retail_data["Price"]

# Generate new column: week
retail_data["week"] = retail_data["InvoiceDate"].dt.to_period("W").apply(lambda r: r.start_time)

transation_value = (
    retail_data.groupby("week").agg(
        sum_value = ("value", "sum"),
        cnt_transation = ("Invoice", "nunique")
    )
)

# Calculate ATV = sum_value / cnt_transation
transation_value["ATV"] = transation_value["sum_value"] / transation_value["cnt_transation"]

#Sorted in DESC
transation_value = transation_value.sort_values(by="ATV", ascending=False)

#Display result
transation_value

sum_value  cnt_transation          ATV
week                                               
2010-12-06  583361.630             532  1096.544417
2010-01-04  168520.110             205   822.049317
2010-11-29  510148.800             657   776.482192
2011-01-10  193362.040             261   740.850728
2011-09-19  338771.791             471   719.260703
...                ...             ...          ...
2011-05-02  146941.600             370   397.139459
2010-05-24  160432.530             406   395.154015
2010-05-31  108579.820             279   389.174982
2010-06-21  127041.530             329   386.144468
2010-02-08   91110.610             281   324.237046

[104 rows x 3 columns]

The week of Dec 6, 2010 has highest ATV, which is 1096.54.

## 3.Which week year has lowest units per transation?

In [ ]:
# Generate new column: week
retail_data["week"] = retail_data["InvoiceDate"].dt.to_period("W").apply(lambda r: r.start_time)

transation_units = (
    retail_data.groupby("week").agg(
        total_units = ("Quantity", "sum"),
        cnt_transation = ("Invoice", "nunique")
    )
)

# Calculate UPT = Total_nuits / cnt_transation
transation_units["UPT"] = transation_units["total_units"] / transation_units["cnt_transation"]

# Sorted by DSC
transation_units = transation_units.sort_values(by="UPT", ascending=True)

# Diplay result
transation_units

total_units  cnt_transation         UPT
week                                               
2010-02-08        51002             281  181.501779
2010-06-21        68663             329  208.702128
2010-07-26        74805             354  211.313559
2011-05-02        78385             370  211.851351
2010-11-22       148492             693  214.274170
...                 ...             ...         ...
2010-03-15       156625             354  442.443503
2010-01-04        96246             205  469.492683
2010-01-18       136829             269  508.657993
2010-08-09       182465             337  541.439169
2010-02-15       176507             313  563.920128

[104 rows x 3 columns]

## 4. Which week has lowest average selling price?

In [39]:
# Generate new column: week
retail_data["week"] = retail_data["InvoiceDate"].dt.to_period("W").apply(lambda r: r.start_time)

retail_data["revenue"] = retail_data["Quantity"] * retail_data["Price"]

sum_revenue = (
    retail_data.groupby("week").agg(
        total_revenue = ("revenue", "sum"),
        total_units = ("Quantity", "sum")
    )
)
# Calcute ASP = Total_Revenue / Total_Units
sum_revenue["ASP"] = sum_revenue["total_revenue"] / sum_revenue["total_units"]
# Sort ASC
sum_revenue = sum_revenue.sort_values(by="ASP", ascending=True)
# Display
sum_revenue

total_revenue  total_units       ASP
week                                            
2010-08-09     180134.320       182465  0.987227
2010-02-15     175148.652       176507  0.992304
2010-01-18     154221.301       136829  1.127110
2010-03-15     190300.321       156625  1.215006
2010-09-27     333403.360       239289  1.393308
...                   ...          ...       ...
2011-01-10     193362.040        89682  2.156085
2011-06-06     220151.720       101980  2.158773
2010-07-26     162355.090        74805  2.170378
2010-12-20      92369.300        40340  2.289769
2010-12-06     583361.630       212146  2.749812

[104 rows x 3 columns]

### Answer
Aug 09, 2010 has the lowest average price, which is 0.987227.